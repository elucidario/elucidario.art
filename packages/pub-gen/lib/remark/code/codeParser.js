import path from "path";
import * as unist from "@elucidario/pkg-unist";
import { parseNodeValue, isPubGenNodeValue } from "../../utils";
import { getPaths, readFile } from "@elucidario/pkg-paths";
import { visit } from "unist-util-visit";
import { Console } from "@elucidario/pkg-console";

const codeParser = (treeOptions) => {
    const { publication, lang, assets, version, pkg, ...extra } = treeOptions;

    const console = new Console(pkg);

    // default docxProcessor to pandoc while we don't implement remark-docx.
    let docxProcessor = "pandoc";
    if (typeof extra.docxProcessor !== "undefined")
        docxProcessor = extra.docxProcessor;

    return (tree, file) => {
        visit(tree, "text", async (node) => {
            if (!isPubGenNodeValue(node.value)) return;

            const { action, filePath, fileOptions } = parseNodeValue(
                node.value,
            );

            if (action !== "code") return;

            const codeData = readFile(
                path.resolve(
                    getPaths().publications,
                    publication,
                    "content",
                    lang,
                    filePath,
                ),
            ).content;

            const { ext } = path.parse(filePath);

            node.type = "paragraph";

            const content = codeData.replace(/\r\n|\r|\n/, "");

            const codeAst = [unist.code(content, ext.replace(".", ""))].filter(
                Boolean,
            );

            node.children = codeAst;
            delete node.value;
        });
    };
};

export default codeParser;
