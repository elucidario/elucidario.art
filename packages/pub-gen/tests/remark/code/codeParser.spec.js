import path from "path";

import codeParser from "../../../lib/remark/code/codeParser";
import { readFile, getPaths } from "@elucidario/pkg-paths";

import testProcessor from "../../mocks/remarkProcessor";

describe("codeParser", () => {
    const codeContent = readFile(
        path.resolve(
            getPaths().publications,
            "publicacao-teste",
            "content",
            "pt-br",
            "nodes",
            "code.md"
        )
    ).content;

    it("should return codeParser node", async () => {
        const codeParserNode = await testProcessor(codeContent, {
            plugin: codeParser,
            pluginOptions: {
                publication: "publicacao-teste",
                lang: "pt-BR",
            },
            stringifyOptions: {},
            publication: "publicacao-teste",
            lang: "pt-BR",
        });

        expect(codeParserNode).toMatchSnapshot();
    });
});
