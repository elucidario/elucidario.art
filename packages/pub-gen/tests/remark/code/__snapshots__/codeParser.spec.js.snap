// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codeParser should return codeParser node 1`] = `
VFile {
  "cwd": "C:\\Users\\55119\\Elucid√°rio.art\\elucidario\\packages\\pub-gen",
  "data": {},
  "history": [],
  "messages": [],
  "value": "\`\`\`js
console.log("code-node.js");
\`\`\`
",
}
`;
